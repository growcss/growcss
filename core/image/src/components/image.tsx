import React, { Component, ReactNode, ReactElement } from 'react';
// eslint-disable-next-line no-unused-vars
import { withTheme, DefaultTheme } from 'styled-components';
// eslint-disable-next-line no-unused-vars
import { getThemeValue } from '@growcss/theme';
import { stripUnit } from '@growcss/elaborate';
import { AspectRatioPlaceholder } from '../styled/aspect-ratio-placeholder';
import { FigureElement } from '../styled/figure-element';
import { PlaceholderElement } from '../styled/placeholder-element';
import { ImageStateType } from '../states';
import { ImageProps, CandidateProps, PictureProps } from '../../types';
import { getDimension } from './helper/get-dimension';
import { parseSrcSet } from './helper/srcset-parser';
import { filterSmall } from './helper/filter-small';
import { filterLarge } from './helper/filter-large';
import { isSsr } from './helper/is-ssr';
import { isNativeConnection } from './helper/is-native-connection';
import { Picture } from './picture';

interface DefaultImageProps {
  onLoad: () => void;
  onError: () => void;
  onStartLoad: () => void;
  onOffline: () => void;
  onSlowNetwork: () => void;
  preload: boolean;
  lazy: boolean;
  webpRegex: RegExp;
  observer: {
    rootMargin: string;
    threshold: number | number[];
  };
}

type PropsWithDefaults = ImageProps & DefaultImageProps;

// Cache if we've seen an image before so we don't bother with
// lazy-loading & fading in on subsequent mounts.
const imageCache: { [key: string]: boolean } = {};

const inImageCache = (src: string): boolean => {
  return imageCache[src] || false;
};

const activateCacheForImage = (src: string): void => {
  imageCache[src] = true;
};

const defaultObserver = {
  rootMargin: '0px 0px 0px 0px',
  threshold: 0,
};

class ExtendedImage extends Component<ImageProps, ImageStateType> {
  public static defaultProps: DefaultImageProps = {
    onLoad: () => {},
    onError: () => {},
    onStartLoad: () => {},
    onOffline: () => {},
    onSlowNetwork: () => {},
    preload: false,
    lazy: false,
    webpRegex: new RegExp('/.+.webp$/i'),
    observer: defaultObserver,
  };

  /**
   * @param {PropsWithDefaults} props
   */
  public constructor(props: PropsWithDefaults) {
    super(props);

    let inViewport = true;

    const {
      src,
      srcSet,
      placeholder,
      theme,
      webpRegex,
      lazy,
    } = props as PropsWithDefaults;
    const srcArray =
      srcSet !== undefined
        ? parseSrcSet(srcSet, getThemeValue('image.breakpoints', {})(theme))
        : [];
    const hasWebpImages =
      srcArray.filter(srcItem => {
        return webpRegex.test(srcItem.url);
      }).length > 0;
    const selectedImage = ExtendedImage.bestMatchingImage(srcArray, src);
    const seenBefore = inImageCache(selectedImage);
    const hasSrcSet = srcArray.length > 1;

    // browser with Intersection Observer available
    if (
      !seenBefore &&
      lazy &&
      typeof window !== `undefined` &&
      window.IntersectionObserver
    ) {
      inViewport = false;
    }

    // Never render image during SSR
    if (typeof window === `undefined` && lazy) {
      inViewport = false;
    }

    this.state = {
      loadState: 'initial',
      possiblySlowNetwork: false,
      inViewport,
      online: true,
      hasWebp: hasWebpImages,
      srcSetWebpArray:
        hasWebpImages && hasSrcSet
          ? ExtendedImage.filterWebp(webpRegex, srcArray, true)
          : undefined,
      srcSetArray: hasSrcSet
        ? ExtendedImage.filterWebp(webpRegex, srcArray, false)
        : undefined,
      placeholder: ExtendedImage.getPlaceholder(srcArray, placeholder),
      src: selectedImage,
      seenBefore,
    };

    this.updateConnection = this.updateConnection.bind(this);
    this.possiblySlowNetworkListener = this.possiblySlowNetworkListener.bind(this);
    this.updateOnlineStatus = this.updateOnlineStatus.bind(this);
    this.onEnter = this.onEnter.bind(this);
  }

  /**
   * {@inheritdoc}
   */
  public componentDidMount(): void {
    const { threshold, onOffline, onSlowNetwork } = this.props as PropsWithDefaults;
    const isOnline = navigator.onLine;

    if (isNativeConnection && isOnline) {
      navigator.connection.addEventListener('onchange', this.updateConnection);
    } else if (threshold !== undefined) {
      window.document.addEventListener(
        'possiblySlowNetwork',
        this.possiblySlowNetworkListener,
      );
    }

    window.addEventListener('online', this.updateOnlineStatus);
    window.addEventListener('offline', this.updateOnlineStatus);

    const { online, possiblySlowNetwork, src } = this.state;

    if (!online && typeof onOffline === 'function') {
      onOffline();
    }

    if (possiblySlowNetwork && typeof onSlowNetwork === 'function') {
      onSlowNetwork();
    }

    activateCacheForImage(src);
  }

  /**
   * {@inheritdoc}
   */
  public componentWillUnmount(): void {
    const { threshold } = this.props;

    if (isNativeConnection) {
      navigator.connection.removeEventListener('onchange', this.updateConnection);
    } else if (threshold !== undefined) {
      window.document.removeEventListener(
        'possiblySlowNetwork',
        this.possiblySlowNetworkListener,
      );
    }

    window.removeEventListener('online', this.updateOnlineStatus);
    window.removeEventListener('offline', this.updateOnlineStatus);
  }

  public onEnter(event, unobserve): void {
    const { inViewport } = this.state;

    if (inViewport) {
      return;
    }

    const { lazy } = this.props;

    if (event.isIntersecting || lazy === false) {
      unobserve();
    }

    this.setState({
      inViewport: !lazy ? true : event.isIntersecting,
    });
  }

  /**
   * {@inheritdoc}
   */
  public render(): ReactNode {
    const { observer, lazy } = this.props as PropsWithDefaults;

    if (lazy) {
      try {
        // eslint-disable-next-line global-require
        const IntersectionObserver = require('@researchgate/react-intersection-observer')
          .default;

        return (
          <IntersectionObserver
            onChange={this.onEnter}
            rootMargin={observer.rootMargin || defaultObserver.rootMargin}
            threshold={observer.threshold || defaultObserver.threshold}
          >
            {this.renderImage()}
          </IntersectionObserver>
        );
      } catch (error) {
        if (error instanceof Error && error.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            'Please run "npm install @researchgate/react-intersection-observer@^1.0.0-beta.0" to use the "lazy" feature of the growcss image component.',
          );
        }
      }
    }

    return this.renderImage();
  }

  /**
   * Update the connection state.
   *
   * @returns {void}
   */
  public updateConnection(): void {
    if (!navigator.onLine) {
      return;
    }

    const { loadState } = this.state;

    if (loadState === 'initial') {
      // @todo use this for a download icon
      /* eslint-disable react/no-unused-state */
      this.setState({
        connection: {
          effectiveType: navigator.connection.effectiveType,
          downlink: navigator.connection.downlink,
          rtt: navigator.connection.rtt,
        },
      });
    }
  }

  /**
   * Update the online state.
   *
   * @returns {void}
   */
  public updateOnlineStatus(): void {
    this.setState({ online: navigator.onLine });
  }

  /**
   * Check if the network is slow and update state.
   *
   * @param {object} erorr
   *
   * @returns void
   */
  public possiblySlowNetworkListener(erorr): void {
    const { loadState, possiblySlowNetwork: statePossiblySlowNetwork } = this.state;

    if (loadState !== 'initial') {
      return;
    }

    const { possiblySlowNetwork } = erorr.detail;

    if (!statePossiblySlowNetwork && possiblySlowNetwork) {
      this.setState({ possiblySlowNetwork });
    }
  }

  /**
   * Render a default placeholder if no placeholder was found.
   *
   * @param {DefaultTheme} theme
   * @param {undefined | string} alt
   * @param {undefined | string} placeholder
   *
   * @returns {ReactNode}
   */
  public static renderPlaceholder(
    theme: DefaultTheme,
    alt?: string,
    placeholder?: string,
  ): ReactNode {
    const defaultSvgPlaceholder = `<svg xmlns='http://www.w3.org/2000/svg' width='1' height='1'><rect width='1' height='1' fill='${getThemeValue(
      'image.backgroundColor',
    )(theme)}'/></svg>`;

    return (
      <PlaceholderElement
        className="placeholder"
        src={placeholder || `data:image/svg+xml;base64,${btoa(defaultSvgPlaceholder)}`}
        crossOrigin="anonymous"
        alt={alt}
      />
    );
  }

  /**
   * @param {string | number | undefined} height
   * @param {string | number | undefined} width
   */
  public static getDimensionRatio(
    height: string | number | undefined,
    width: string | number | undefined,
  ): number {
    if (height !== undefined && width !== undefined) {
      return (stripUnit(height) / stripUnit(width)) * 100;
    }

    const { height: dHeight, width: dWidth } = getDimension();

    return (dHeight / dWidth) * 100;
  }

  /**
   * Get best image from srcset
   *
   * @param {CandidateProps[]} srcSetArray
   * @param {undefined|string} src
   *
   * @returns {string}
   */
  public static bestMatchingImage(srcSetArray: CandidateProps[], src?: string): string {
    let densityMultiplier = 1;

    // Check if denisty is provided in srcset
    let hasDensity = false;

    srcSetArray.forEach(srcSet => {
      if ('density' in srcSet) {
        hasDensity = true;
      }
    });

    if (!hasDensity) {
      // When no density is set use devicePixelRatio as a multiplier for reference dimensions
      densityMultiplier = window.devicePixelRatio || 1.0;
    }

    const refDim: { width: number; height: number } = getDimension();
    let filtered: CandidateProps[] = srcSetArray;

    if (typeof src === 'string' && src !== '') {
      filtered = srcSetArray.concat([
        {
          url: src,
        },
      ]);
    }

    filtered = filterLarge(
      filterLarge(
        filterLarge(filtered, 'width', refDim.width * densityMultiplier),
        'height',
        refDim.height * densityMultiplier,
      ),
      'density',
      !hasDensity ? 1 : window.devicePixelRatio || 1.0, // When no density is provided densityMultiplier is used and density filter can be set to 1, otherwise use devicePixelRatio
    );

    filtered = filterSmall(
      filterSmall(filterSmall(filtered, 'width'), 'height'),
      'density',
    );

    return filtered[0].url;
  }

  /**
   * Returns a filtered copy of the source set array
   * including only webp images if supports is true,
   * or only non webp images if supports is false.
   *
   * @param {RegExp}           webpRegex
   * @param {CandidateProps[]} srcArray
   * @param {boolean}          supports Indicates if the browser supports webp.
   *
   * @returns CandidateProps[] filtered source set array.
   */
  public static filterWebp(
    webpRegex: RegExp,
    srcArray: CandidateProps[],
    supports: boolean,
  ): CandidateProps[] {
    return srcArray.filter(srcItem => {
      return webpRegex.test(srcItem.url) === supports;
    });
  }

  /**
   *
   * @param {CandidateProps[]} srcArray
   * @param {undefined | string} placeholder
   */
  public static getPlaceholder(
    srcArray: CandidateProps[],
    placeholder?: string,
  ): undefined | string {
    if (placeholder === undefined && srcArray.length !== 0) {
      const smallImages = filterSmall(
        filterSmall(filterSmall(srcArray, 'width'), 'height'),
        'density',
      );

      return smallImages.length !== 0 ? smallImages[0].url : undefined;
    }

    return placeholder;
  }

  /**
   * Render Image component.
   *
   * @returns {ReactElement<PictureProps>}
   */
  private renderImage(): ReactElement<PictureProps> {
    const {
      children,
      height,
      width,
      preload,
      theme,
      onError,
      onLoad,
      onStartLoad,
      alt,
      sizes,
      crossOrigin,
      ...other
    } = this.props as PropsWithDefaults;

    const {
      src,
      srcSetWebpArray,
      srcSetArray,
      placeholder,
      hasWebp,
      loadState,
      inViewport,
      seenBefore,
    } = this.state;

    const fadeIn = preload && !seenBefore;

    return (
      <FigureElement className="gc-image">
        <AspectRatioPlaceholder>
          <div
            style={{
              width: '100%',
              paddingBottom: `${ExtendedImage.getDimensionRatio(height, width)}%`,
            }}
          />
          {fadeIn && ExtendedImage.renderPlaceholder(theme, alt, placeholder)}
          {inViewport && (
            <Picture
              src={src}
              state={loadState}
              preload={preload}
              srcSetArray={srcSetArray}
              srcSetWebpArray={srcSetWebpArray}
              hasWebp={hasWebp}
              onLoad={onLoad}
              onError={onError}
              onStartLoad={onStartLoad}
              crossOrigin={crossOrigin}
              sizes={sizes}
              alt={alt}
              {...other}
            />
          )}
          {isSsr && (
            <noscript>
              <Picture
                src={src}
                state="loaded"
                preload={false}
                srcSetArray={srcSetArray}
                srcSetWebpArray={srcSetWebpArray}
                hasWebp
                crossOrigin={crossOrigin}
                sizes={sizes}
                alt={alt}
                {...other}
              />
            </noscript>
          )}
        </AspectRatioPlaceholder>
        {children}
      </FigureElement>
    );
  }
}

export const StyledImage = withTheme(ExtendedImage);
