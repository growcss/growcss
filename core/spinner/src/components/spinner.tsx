import React, { Component, ReactNode, ComponentClass, ClassAttributes } from 'react';
// eslint-disable-next-line no-unused-vars
import { AnimationProps, SpinnerProps, SpinnerState } from '../../types';
import { FingerprintAnimation } from './animations/fingerprint-animation';
import { TwoHalfCircleAnimation } from './animations/two-half-circle-animation';
import { HalfCircleAnimation } from './animations/half-circle-animation';
import { SpinnerElement } from './styled/spinner-element';

/**
 * A list of named sizes for spinner.
 *
 * @type {{small: number, medium: number, large: number, xlarge: number, xxlarge: number}}
 */
export const Size: { [key: string]: number } = {
  small: 8,
  medium: 16,
  large: 24,
  xlarge: 48,
  xxlarge: 96,
};

interface DefaultSpinnerProps extends SpinnerProps {
  delay: number;
  size: string | number;
  animation: string;
  isCompleting: boolean;
}

export class Spinner extends Component<SpinnerProps, SpinnerState> {
  public static defaultProps: DefaultSpinnerProps = {
    delay: 10,
    size: 'large',
    animation: 'halfCircle',
    color: '#000',
    isCompleting: false,
  };

  public render(): ReactNode {
    const { delay, animation, color, isCompleting } = this.props as DefaultSpinnerProps;
    const animations: { [key: string]: ComponentClass } = {
      fingerprint: FingerprintAnimation,
      twoHalfCircle: TwoHalfCircleAnimation,
      halfCircle: HalfCircleAnimation,
    };
    const size = this.getSize();

    return (
      <SpinnerElement className="gc-spinner">
        {React.createElement(
          animations[animation] as ComponentClass,
          // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
          {
            size,
            delay,
            color,
            isCompleting,
          } as ClassAttributes<AnimationProps>,
        )}
      </SpinnerElement>
    );
  }

  /**
   * Returns spinner size number.
   *
   * @returns number
   */
  private getSize(): number {
    const { size } = this.props as DefaultSpinnerProps;
    const spinnerSize = Size[size] || size;

    return typeof spinnerSize === 'number' ? spinnerSize : Size.small;
  }
}
