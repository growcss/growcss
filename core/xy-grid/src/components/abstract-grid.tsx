import React, { PureComponent, ReactNode, ReactElement, Children } from 'react';
import classNames from 'classnames';
import { Gutters as DefaultGutters } from './gutters';
import { XYGridElement } from '../styled/xy-grid-element';
// eslint-disable-next-line no-unused-vars
import { GridProps, CellProps, GuttersProps } from '../../types';

interface DefaultGridProps {
  gutterSizes?: GuttersProps;
  alignX?: string;
}

type PropsWithDefaults = GridProps & DefaultGridProps;

export default class AbstractGrid extends PureComponent<GridProps> {
  public static defaultProps: DefaultGridProps = {
    gutterSizes: DefaultGutters,
    alignX: 'left',
  };

  /**
   * Setting for getting a gridy or gridx type.
   *
   * @param {boolean} vertical
   */
  protected vertical: boolean;

  /**
   * @inheritdoc
   */
  public render(): ReactNode {
    const { children, gutterSizes, gutterType, alignX, alignY, height, ...other } = this
      .props as PropsWithDefaults;
    const className = classNames(`gc-grid-${this.vertical ? 'y' : 'x'}`);
    const direction = this.vertical ? 'vertical' : 'horizontal';

    if (this.vertical && height === undefined) {
      throw new Error(
        'Y-Grid needs always a height, please use "height" prop to add the grid height.',
      );
    }

    return (
      <XYGridElement
        className={className}
        gridDirection={direction}
        alignX={alignX}
        alignY={alignY}
        gridHeight={height}
        flexWrap
        {...other}
      >
        {this.renderCellChildren(children, gutterType, gutterSizes)}
      </XYGridElement>
    );
  }

  /**
   * Renders the child elements.
   *
   * @param {Component<CellProps>[]}   children
   * @param {undefined | string}       gutterType
   * @param {undefined | GuttersProps} gutterSizes
   *
   * @returns {ReactNode[]}
   */
  private renderCellChildren(
    children: ReactNode,
    gutterType?: string,
    gutterSizes?: GuttersProps,
  ): ReactNode[] {
    return Children.map(children, (thisArg: ReactElement<CellProps>) => {
      return React.cloneElement(thisArg, {
        gutterType,
        gutterSizes,
        vertical: this.vertical,
      });
    });
  }
}
