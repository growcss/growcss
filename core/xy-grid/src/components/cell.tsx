import React, { PureComponent, ReactNode } from 'react';
import classNames from 'classnames';
import { Gutters as DefaultGutters } from './gutters';
import { CellElement } from '../styled/cell-element';
// eslint-disable-next-line no-unused-vars
import { CellBreakpointsProps, CellProps, GuttersProps, OffsetProps } from '../../types';

interface DefaultCellProps {
  gridColumns: number;
  gutterSizes: GuttersProps;
  vertical: boolean;
}

type PropsWithDefaults = CellProps & DefaultCellProps;

export class Cell extends PureComponent<CellProps & CellBreakpointsProps & OffsetProps> {
  public static defaultProps: DefaultCellProps = {
    gridColumns: 12,
    gutterSizes: DefaultGutters,
    vertical: false,
  };

  public render(): ReactNode {
    const {
      children,
      gridColumns,
      cellType,
      gutterType,
      gutterSizes,
      vertical,
      align,
      ...other
    } = this.props as PropsWithDefaults;
    const className = classNames('gc-cell');

    return (
      <CellElement
        className={className}
        gridColumns={gridColumns}
        cellType={cellType}
        gutterType={gutterType}
        gutterSizes={gutterSizes}
        vertical={vertical}
        align={align}
        {...other}
      >
        {children}
      </CellElement>
    );
  }
}
