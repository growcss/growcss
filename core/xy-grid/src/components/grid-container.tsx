import React, { PureComponent, ReactNode } from 'react';
import classNames from 'classnames';
import { rem } from '@growcss/elaborate';
import { Gutters as DefaultGutters } from './gutters';
import { GridContainerElement } from '../styled/grid-container-element';
import { GridContainerProps, GuttersProps } from '../../types';

interface DefaultGridContainerProps {
  width: string;
  gutterSizes: GuttersProps;
}

type PropsWithDefaults = GridContainerProps & DefaultGridContainerProps;

export class GridContainer extends PureComponent<GridContainerProps> {
  public static defaultProps: DefaultGridContainerProps = {
    width: rem(1200),
    gutterSizes: DefaultGutters,
  };

  public render(): ReactNode {
    const { type, children, width, gutterSizes, ...other } = this
      .props as PropsWithDefaults;
    const className = classNames('gc-grid-container');

    let maxWidth = width;
    let gutter: string | number | GuttersProps = gutterSizes;

    if (type === 'fluid' || type === 'full') {
      maxWidth = '100%';
    }

    if (type === 'full') {
      gutter = '0';
    }

    return (
      <GridContainerElement
        className={className}
        maxWidth={maxWidth}
        gutterSizes={gutter}
        type={type}
        {...other}
      >
        {children}
      </GridContainerElement>
    );
  }
}
