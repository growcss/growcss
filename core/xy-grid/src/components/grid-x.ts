import AbstractGrid from './abstract-grid';

export class GridX extends AbstractGrid {
  /**
   * @inheritdoc
   */
  protected vertical: boolean = false;
}
