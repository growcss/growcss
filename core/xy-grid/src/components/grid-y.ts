import AbstractGrid from './abstract-grid';

export class GridY extends AbstractGrid {
  /**
   * @inheritdoc
   */
  protected vertical: boolean = true;
}
