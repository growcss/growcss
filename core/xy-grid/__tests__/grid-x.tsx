import * as React from 'react';
import { mount } from 'enzyme';
//import renderer from 'react-test-renderer';
import { GridX } from '../src/components/grid-x';
import { Cell } from '../src/components/cell';

test('the style output of GridX', () => {
  const grid = (
    <GridX>
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // @todo waiting for enzyme fix
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('justify-content', 'flex-start');
});

test('the style output of GridX with alignment to right', () => {
  const grid = (
    <GridX gutterType="padding" alignX="right">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('justify-content', 'flex-end');
});

test('the style output of GridX with alignment to center', () => {
  const grid = (
    <GridX gutterType="padding" alignX="center">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('justify-content', 'center');
});

test('the style output of GridX with alignment to justify', () => {
  const grid = (
    <GridX gutterType="padding" alignX="justify">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('justify-content', 'space-between');
});

test('the style output of GridX with alignment to spaced', () => {
  const grid = (
    <GridX gutterType="padding" alignX="spaced">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('justify-content', 'space-around');
});

test('the style output of GridX with alignment to top', () => {
  const grid = (
    <GridX gutterType="padding" alignY="top">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('align-items', 'flex-start');
});

test('the style output of GridX with alignment to middle', () => {
  const grid = (
    <GridX gutterType="padding" alignY="middle">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('align-items', 'center');
});

test('the style output of GridX with alignment to bottom', () => {
  const grid = (
    <GridX gutterType="padding" alignY="bottom">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('align-items', 'flex-end');
});

test('the style output of GridX with alignment to stretch', () => {
  const grid = (
    <GridX gutterType="padding" alignY="stretch">
      <Cell>test</Cell>
    </GridX>
  );
  // const tree = renderer.create(grid).toJSON();

  expect(mount(grid)).toMatchSnapshot();
  // expect(tree).toHaveStyleRule('display', 'flex');
  // expect(tree).toHaveStyleRule('flex-flow', 'row wrap');
  // expect(tree).toHaveStyleRule('-webkit-box-orient', 'horizontal');
  // expect(tree).toHaveStyleRule('-webkit-box-direction', 'normal');
  // expect(tree).toHaveStyleRule('align-items', 'stretch');
});
