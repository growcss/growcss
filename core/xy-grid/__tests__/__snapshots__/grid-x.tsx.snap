// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the style output of GridX 1`] = `
<xy-grid-element__XYGridElement
  alignX="left"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 jqHyhp"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 PmlGH"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to bottom 1`] = `
<xy-grid-element__XYGridElement
  alignX="left"
  alignY="bottom"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 fYSDDe"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to center 1`] = `
<xy-grid-element__XYGridElement
  alignX="center"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 gSAUKy"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to justify 1`] = `
<xy-grid-element__XYGridElement
  alignX="justify"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 imMcgZ"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to middle 1`] = `
<xy-grid-element__XYGridElement
  alignX="left"
  alignY="middle"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 beHfpv"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to right 1`] = `
<xy-grid-element__XYGridElement
  alignX="right"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 lbjrIu"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to spaced 1`] = `
<xy-grid-element__XYGridElement
  alignX="spaced"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 gwaiGx"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to stretch 1`] = `
<xy-grid-element__XYGridElement
  alignX="left"
  alignY="stretch"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 flrYIA"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;

exports[`the style output of GridX with alignment to top 1`] = `
<xy-grid-element__XYGridElement
  alignX="left"
  alignY="top"
  className="gc-grid-x"
  flexWrap={true}
  gridDirection="horizontal"
>
  <div
    className="gc-grid-x xy-grid-element__XYGridElement-sc-76t78x-0 kbQcJl"
  >
    <cell-element__CellElement
      className="gc-cell"
      gridColumns={12}
      gutterSizes={
        Object {
          "medium": "30px",
          "small": "20px",
        }
      }
      gutterType="padding"
      vertical={false}
    >
      <div
        className="gc-cell cell-element__CellElement-sc-6s0nu0-0 eRyEvK"
      >
        test
      </div>
    </cell-element__CellElement>
  </div>
</xy-grid-element__XYGridElement>
`;
